### PR TITLE
#27 Vector Implementation

### PR DESCRIPTION
**Pull request by Oggeboii**
Pr Issue #27 Vector Implementation

## Summary by CodeRabbit

- **New Features**
  - Introduced versatile vector types for 2D, 3D, and 4D space with comprehensive arithmetic operations, including addition, subtraction, multiplication, and division.
  - Enabled key vector functions such as dot product, cross product (for 3D), magnitude, normalization, Euclidean distance, linear interpolation, and component-wise comparisons.
  - Provided seamless conversion methods among different vector dimensions.

- **Tests**
  - Implemented extensive tests for `Vector2`, `Vector3`, and `Vector4` to ensure accuracy, proper error handling, and stable performance across all vector operations.

### CODE REVIEW

**Review by CarlGullberg — DISMISSED** [review]
Looks good, great descriptions on the tests!!

Reviewed with @AntonGrubisic

**Review by AntonGrubisic — APPROVED** [review]
Resolved changes made by AI. Code looks good.

Reviewed with @CarlGullberg

File: src/test/java/org/fungover/breeze/vector/Vector2Test.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        );
+    }
+
+    @Test
+    @DisplayName("Correct dot product")
+    void correctDotProduct() {
+        Vector2 vector = new Vector2(3, 3);
+        Vector2 vector2 = new Vector2(3, 3);
+        assertThat(vector.dot(vector2)).isEqualTo(18.0f);
+    }
```
Comment (tovabry) [inline src/test/java/org/fungover/breeze/vector/Vector2Test.java:65]: A more informative name could be Should return correct dot product or dot method should return correct dot product?

**Review by tovabry — COMMENTED** [review]
This is a very complicated issue that you seem to have managed with bravura. And here I thought 4D was about smell or something... 
You seem to use a minimalistic naming style both in the variables in the vector classes and the tests wich surely has a purpose - but could some of the tests be more informative?

**Review by tovabry — APPROVED** [review]
Nice changes!

**Review by tovabry — APPROVED** [review]

**Review by tovabry — APPROVED** [review]

**Review by CarlGullberg — APPROVED** [review]
Looks good, great discription for everything. Well done!

**Review by tovabry — APPROVED** [review]

**Comment by kappsegla** [comment]
So this implementation is a mutable implementation that changes the actual object when calling a method on the vector classes. Good for performance in a game where don't have to create lots of new objects. But it would also mean that the risk of unintentional changes increases.

The alternative is to do a separate immutable implementation?

Coderabbit had some suggestions 4 days ago about changing the mutation methods to return new instances of the classes. It is marked as resolved by @AntonGrubisic without any comments or changes done to the code?

**Review by CarlGullberg — APPROVED** [review]
Looks good after the recent changes.

Reviewed with @AntonGrubisic

**Review by tovabry — DISMISSED** [review]

File: src/test/java/org/fungover/breeze/vector/Vector4Test.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+
+class Vector4Test {
+
+    @Test
+    @DisplayName("Add method adds new values to old values")
+    void addMethodAddsNewValuesToOldValues() {
```
Comment (Minhya) [inline src/test/java/org/fungover/breeze/vector/Vector4Test.java:15]: Great tests in general, however consider adding tests where vectors have both positive and negative components or are not symmetric. This ensure that nothing goes wrong in those cases :)

**Review by kappsegla — APPROVED** [review]

**Review by AntonGrubisic — APPROVED** [review]
Perfect!
